OpenSim/Sandbox/DataFormattersAndAdapters.h

```diff
@@ -0,0 +1,212 @@
+public:
+    OutputTables read() const;
+    void write(InputTables);
+    virtual ~DataAdapter() = default;
+
+protected:
+    DataAdapter() = default;
+    virtual OutputTables implementRead() const = 0;
+    virtual void implementWrite(InputTables) = 0;
+};
+
+/** This is a DataAdapter for use with an already-existing read/write stream.
+For read-only streams (istream) like std::cin or write-only (ostream) like
+std::cout we'll need IStreamDataAdapter and OStreamDataAdapter also. Those
+should raise exceptions if disallowed read() or write() is called.
+**/
+class StreamDataAdapter : public DataAdapter {
+public:
+    /** Associate this adapter with a stream and a format for that stream. It
+    is an error if the formatKey cannot be found in the DataStreamFormatter
+    registry. **/
+    StreamDataAdapter(iostream& stream, const std::string& formatKey) 
+    :   m_stream(stream) {
+        m_formatter = DataStreamFormatter::createFormatter(formatKey);
+    }
+private:
+    OutputTables implementRead() const override 
+    {   return m_formatter->read(m_stream); }
+    void implementWrite(InputTables input) override 
+    {   return m_formatter->write(m_stream, input); }
+
+    std::iostream&                  m_stream;
+    unique_ptr<DataStreamFormatter> m_formatter;
+};
+
+/** A string adapter is a kind of stream adapter. **/
+class StringDataAdapter : public StreamDataAdapter {
+public:
+    StringDataAdapter(const std::string& data, const std::string& formatKey)
+    :   StreamDataAdapter(std::stringstream(data), formatKey) {}
+};
+
+/** Helper class for when you know this file will be handled using one of
+the DataStreamFormatters. Users will use the FileDataAdapter instead. **/
+class FileStreamDataAdapter : public StreamDataAdapter {
+public:
+    FileStreamDataAdapter(const std::string& pathName)
+    :   StreamDataAdapter(std::fstream(pathName), 
+                          findExtension(pathName)) {}
+};
+
+/** Helper class for when you know this file will be handled using the
+C3dFormatter. Users will use the FileDataAdapter instead. **/
+class C3dFileDataAdapter : public DataAdapter {
+public:
+    C3dFileDataAdapter(const std::string& c3dFileName) 
+    :   m_formatter(c3dFilesName) {}
+private:
+    OutputTables implementRead() const override 
+    {   return m_formatter->read(m_stream); }
+    void implementWrite(InputTables input) override 
+    {   return m_formatter->write(m_stream, input); }
+
+    C3dFormatter m_formatter;
+};
+
+
+/** Instantiate a DataAdapter for a particular file given by name. This class 
+figures out the appropriate format to use for the file. **/
+class FileDataAdapter : public DataAdapter {
+public:
+    FileDataAdapter(const std::string& pathName) {
+        const std::string ext = findExtension(pathName);
+        if (ext == "c3d") {
+            m_adapter.reset(new C3dFileDataAdapter(pathName));
+        } else
+            m_adapter.reset(new FileStreamDataAdapter(pathName));
```

Comment:
How do I register a data format for MAT or HDF5? I probably don't want to use `FileStreamDataAdapter`, and I think we should expect third parties will want to write support for formats that opensim itself does not support.